prober/icmp.go

```diff
@@ -131,6 +137,7 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
 			}
 		}
 
+		icmpConn.IPv6PacketConn().SetControlMessage(ipv6.FlagHopLimit, true)
```

Comment:
Are we changing the outbound packet here? The docs don't indicate what this actually does.

Reply:
SetControlMessage asks for extra info when a datagram is received (inside controlMessage). It might be not fully documented inside go because it is not something special for it. It is a UNIX thing.

> ...
> The  field  msg_control, which has length msg_controllen, points to **a buffer for other protocol control-related messages or miscellaneous ancillary data**.  When recvmsg() is called, msg_controllen should contain the length of the available buffer in msg_control; upon return from a successful call it will contain the length of the control message sequence.
> ...
> recvmsg(2) manual

> ...
> IP_RECVTTL (since Linux 2.2)
> When this flag is set, pass a IP_TTL control message with the time-to-live field of the received packet as a 32 bit integer.  Not supported for SOCK_STREAM sockets.
> ...
> ip(7) manual

> ...
>IPV6_RTHDR, IPV6_AUTHHDR, IPV6_DSTOPTS, IPV6_HOPOPTS, IPV6_FLOWINFO, **IPV6_HOPLIMIT**
>Set delivery of control messages for incoming datagrams containing extension headers from the received packet.  IPV6_RTHDR delivers the routing header, IPV6_AUTHHDR delivers the authentication header, IPV6_DSTOPTS delivers the destination options, IPV6_HOPOPTS delivers the hop options, IPV6_FLOWINFO delivers an integer containing the flow ID, **IPV6_HOPLIMIT delivers an integer containing the hop count of the packet**. The  control messages have the  same type as the socket option. All these header options can also be set for outgoing packets by putting the appropriate control message into the control buffer of sendmsg(2). Allowed only for SOCK_DGRAM or SOCK_RAW sockets.  Argument is a pointer to a boolean value.
> ...
> ipv6(7) manual

AFAIK, at least for those flags, it should not affect outgoing datagrams, even only checking go docs:

> TTL     int    // time-to-live, **receiving only**
https://github.com/golang/net/blob/master/ipv4/control.go#L35
> FlagHopLimit                              // pass the hop limit on the **received packet**
https://github.com/golang/net/blob/master/ipv6/control.go#L32

It might not work with Windows machines. I don't have a windows machine to test Prometheus but it will be 0 if missing (hoplimit 0 will never naturally occur as it will get dropped before reaching us).

How to deal with missing data in Prometheus? Null?? Hide the property? Use zero? -1?

Reply:
The question is more will setting this break icmp on Windows. This will need to be tested manually.

Reply:
For Windows it will (currently) return a NotImplemented error (https://github.com/golang/net/blob/5d4f7005572804eaf7f5ecdd2473a62557f733ba/ipv4/control_windows.go#L11), although there is a "// TODO(mikio): implement this". It might need to map UNIX specific code to Windows.

When NotImplemented is returned, my code will write a msg to log. Nothing else will change.

The cm was already being "returned" but ignored (the "_" here https://github.com/prometheus/blackbox_exporter/pull/694/files#diff-57e28fd17986190d6a1ba451a5131874L291

The cm is initialized inside packetHandler.ReadFrom() if message.NN is positive (The NN field indicates the number of bytes read or written from/to OOB). For windows, I'm not sure if OOB is empty or not. So, I need to consider it might be nil.
I added an if before using it.

Now, it might not break anything even in Windows. I just don't have a Windows to test.

The question that remains is about missing data. According to [this](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics), I should give a metric even if missing. I could use NaN or 0. If I do nothing, it will be 0. ipv4 TTL==0 is not valid even for localhost while IPv6 HopLimit could be 0 between hosts in same network. As I'm getting it from an ICMP echo-reply, HopLimit == 0 might be valid when the answer reaches the probe exactly one hop from being dropped. It might be too rare to bother. Someone might manually set echo-request HopLimit/TTL to count hops but not echo-response HopLimit/TTL (which is what I'm monitoring). Other metrics, like probe_icmp_duration_seconds, give me 0 when they fail, even when 0 is a "valid duration". NaN seems to be better to avoid "missing/ipv6 last hop" dubious meaning. However, it would make sense if all other metrics also had the same semantic. I guess changing a metric semantic (0 => NaN) is dangerous as it might break someone consuming it.

Reply:
The right way to do it would be not to expose it when it's missing, how the blackbox exporter generally handles this predates staleness. NaN shouldn't be used here.

Reply:
OK, I'll register the gauge only when setControlMessage works, and controlMessage is not nil.